### PR TITLE
Add flattenedToList and flattenedToSet

### PR DESCRIPTION
**Pull request by natebosch**
For iterables which are known to be exhausted after flattening
performance is better than the `sync*` implementation using a collection
literal.

Add `flattenedToList` as a performance improvement over `flattened.`
Add `flattenedToSet` as new behavior for flattening to unique elements.

Originally implemented in
https://github.com/dart-lang/sdk/commit/8d3b6ce54ccb09dd970cd45fc3cded4a35beac31

### CODE REVIEW

**Comment by natebosch** [comment]
cc @scheglov - any concerns about landing this here? You might need to handle import conflicts when this version gets pulled in to the SDK.

**Comment by natebosch** [comment]
I'm not sure if we want to hold this on https://github.com/dart-lang/collection/pull/284 - we also have another new API on that branch waiting for publish.

File: test/extensions_test.dart
Comment (lrhn) [inline test/extensions_test.dart:1084]: Also test that duplicates are removed.

**Review by lrhn — APPROVED** [review]

**Comment by lrhn** [comment]
If someone needs this, it's fine to release it in 2.X, and merge it into the 3.0 branch too.
Otherwise it's probably easier to just put it into the next major release.

I really should get back to releasing the 3.0 version.

**Comment by scheglov** [comment]
> cc @scheglov - any concerns about landing this here? You might need to handle import conflicts when this version gets pulled in to the SDK.

https://dart-review.googlesource.com/c/sdk/+/346163 renames these methods in the analyzer, so there will be no conflict. I will migrate once this PR lands, and `package:collection` is pulled into SDK.

**Comment by kevmoo** [comment]
@natebosch – care to proceed?

File: test/extensions_test.dart
Comment (kevmoo) [reply test/extensions_test.dart:1114]: done!

**Pull request merged**
